mozilla_django_oidc/middleware.py

```diff
@@ -179,3 +189,59 @@ def process_request(self, request):
             response['refresh_url'] = redirect_url
             return response
         return HttpResponseRedirect(redirect_url)
+
+
+class RefreshOIDCAccessToken(SessionRefresh):
+    """
+    A middleware that will refresh the access token following proper OIDC protocol:
+    https://auth0.com/docs/tokens/refresh-token/current
+    """
+    def process_request(self, request):
+        if not self.is_expired(request):
+            return
+
+        token_url = import_from_settings('OIDC_OP_TOKEN_ENDPOINT')
+        client_id = import_from_settings('OIDC_RP_CLIENT_ID')
+        client_secret = import_from_settings('OIDC_RP_CLIENT_SECRET')
+        refresh_token = request.session.get('oidc_refresh_token')
+        if not refresh_token:
+            LOGGER.debug('no refresh token stored')
+            return
+
+        token_payload = {
+            'grant_type': 'refresh_token',
+            'client_id': client_id,
+            'client_secret': client_secret,
+            'refresh_token': refresh_token,
+        }
+
+        try:
+            response = requests.post(
+                token_url,
+                data=token_payload,
+                verify=import_from_settings('OIDC_VERIFY_SSL', True)
+            )
+            response.raise_for_status()
+            token_info = response.json()
+        except requests.exceptions.Timeout:
+            LOGGER.debug('timed out refreshing access token')
+            return
+        except requests.exceptions.HTTPError as exc:
+            LOGGER.debug('http error %s when refreshing access token',
+                exc.response.status_code)
+            return
```

Comment:
1 - Maybe the user should be logged out at this point if the status_code == 401
2 - An appropriate response should be sent in case the request is an ajax call